source/common/http/filter_manager.h

```diff
@@ -837,8 +836,20 @@ class FilterManager : public ScopeTrackedObject,
   virtual StreamInfo::StreamInfo& streamInfo() PURE;
   virtual const StreamInfo::StreamInfo& streamInfo() const PURE;
 
-  // Set up the Encoder/Decoder filter chain.
-  bool createFilterChain();
+  struct CreateFilterChainResult {
+    bool created{};
+    bool upgrade_rejected{};
+  };
+
+  /**
+   * Set up the Encoder/Decoder filter chain.
+   * @param filter_chain_factory the factory to create the filter chain.
+   * @param only_create_if_configured whether to only create the filter chain if it is configured
+   *        explicitly. This only makes sense for upstream HTTP filter chain.
```

Comment:
ooc is it ever not true for upstream filter chain?  I'm wondering if we can infer this as we do the upgrade allowed option if they're just the inverse of each other

Reply:
> ooc is it ever not true for upstream filter chain?

Yeah...it will be false when we try to create filters from router and cluster at second time. As I said before, i think we should remove this by providing a default filter chain factory for upstream filter chain.

Reply:
Would it be worth doing that now?  default filter chain is just codec filter right?

Reply:
> Would it be worth doing that now?  default filter chain is just codec filter right?

I think it deserves. It not just make code simple also reduce memory overhead of cluster because we now need keep a default filter chain for every cluster.